src/packages/BooleanQueryTokenizer/BooleanQueryTokenizer.ts

```diff
@@ -148,6 +155,94 @@ export class BooleanQueryTokenizer {
         );
     }
 
+    public static extractWords(inputString: string): Extract {
+        const tokens = BooleanQueryTokenizer.tokenize(inputString);
+        for (let i = 0; i < tokens.length; i += 1) {
+            const token = tokens[i];
+            if (BooleanQueryTokenizer.isNegation(token) && i + 1 < tokens.length) {
+                let j = i + 1;
+                while (j < tokens.length) {
+                    const lookAheadToken = tokens[j];
+                    if (lookAheadToken.type === TokenType.separator) {
+                        j += 1;
+                    } else if (BooleanQueryTokenizer.isNegationable(lookAheadToken)) {
+                        BooleanQueryTokenizer.replaceTokens(tokens, i, j);
+                        break;
+                    } else if (BooleanQueryTokenizer.isParenthesisGroup(lookAheadToken)) {
+                        let parenthesisTracker = 0;
+                        for (let k = j; k < tokens.length; k += 1) {
+                            const parenthesisLoopToken = tokens[k];
+                            if (parenthesisLoopToken.type === TokenType.parentheses) {
+                                if (parenthesisLoopToken.lexeme === '(') {
+                                    parenthesisTracker += 1;
+                                } else if (parenthesisLoopToken.lexeme === ')') {
+                                    parenthesisTracker -= 1;
+                                }
+                            }
+                            if (parenthesisTracker === 0) {
+                                BooleanQueryTokenizer.replaceTokens(tokens, i, k);
```

Comment:
I think at this moment we also can do `i = k`, because we do need to go through reseted tokens again in main loop.

The same in line 169 we can de `i = j`

Reply:
> because we do need to go through

you mean "we do NOT need to go... ?"

Reply:
nice optimisation anyways :)

Reply:
> you mean "we do NOT need to go... ?"

yes =)